### PR TITLE
Read/Write Abstraction

### PR DESCRIPTION
**Pull request by kripken**
This implements ModuleReader/Writer classes, that can read/write modules in either binary or text form. They pick which according to the suffix, or if the suffix is ambiguous, they default to text.

### CODE REVIEW

**Comment by dschuff** [comment]
I don't think I really like the idea of letting the suffix always determine the file type; I think it will just lead to user surprise if the tools don't behave the same way given the same flags. I think the tools should default to writing binary (except `wasm-dis` of course) and have a flag (`-S`?) to control the output type. For reading we could just always sniff for binary and fall back to text (although that might lead to problematic error messages if the fiile is neither). That would mean that the generic write method could take a flag to specify the type, or just expose 2 different methods.

@sunfishcode you mentioned you were looking at the `s2wasm`-alike that would read binary object files but that the read/write abstraction wasn't right; Is this what you wanted?

**Comment by kripken** [comment]
I guess I prefer the suffix, but you probably have a better idea for what is consistent with other tools. If no one else disagrees then let's go with your way.

**Comment by dschuff** [comment]
I don't know of any toolchain-y tool that changes its output format based on filename (as soon as I say that, I'm sure someone will prove me wrong....). Some change their input behavior (e.g. clang/gcc if not overriden by a flag) but in our case the formats are different enough that sniffing should be no problem.

**Comment by kripken** [comment]
Updated to do it that way, no influence from the suffix.

File: check.py [hunk 1]
```diff
@@ -57,10 +57,20 @@
...
 run_command(cmd)
 fail_if_not_identical(open('a.wast').read(), open(wast).read())
 
+print '\n[ checking wasm-opt binary reading/writing... ]\n'
+
+shutil.copyfile(os.path.join(options.binaryen_test, 'hello_world.wast'), 'a.wast')
+delete_from_orbit('a.wasm')
+delete_from_orbit('b.wast')
+run_command(WASM_OPT + ['a.wast', '-o', 'a.wasm'])
+assert open('a.wasm', 'rb').read()[0] == '\0', '.wasm output suffix means we emit binary'
```
Comment (dschuff) [inline check.py]: Could change these messages to reflect command line now.

Comment (kripken) [inline check.py]: Thanks, done.

File: src/wasm/wasm-io.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+void ModuleReader::readBinary(std::string filename, Module& wasm) {
+  if (debug) std::cerr << "reading binary from " << filename << "\n";
+  auto input(read_file<std::vector<char>>(filename, Flags::Binary, debug ? Flags::Debug : Flags::Release));
+  WasmBinaryBuilder parser(wasm, input, debug);
+  parser.read();
+}
+
+void ModuleReader::read(std::string filename, Module& wasm) {
+  // see if this is a binary
+  auto contents = read_file<std::vector<char>>(filename, Flags::Binary, debug ? Flags::Debug : Flags::Release);
```
Comment (dschuff) [inline src/wasm/wasm-io.cpp]: I just realized; does this read the entire file just to sniff the header? maybe we should fix that.

Comment (kripken) [inline src/wasm/wasm-io.cpp]: Good point, yeah: #894.

**Pull request merged**